Mathlib/Topology/UniformSpace/CoveringProfinite.lean

```diff
@@ -0,0 +1,115 @@
+/-
+Copyright (c) 2025 David Loeffler. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: David Loeffler
+-/
+import Mathlib.Topology.Algebra.Indicator
+import Mathlib.Topology.Separation.DisjointCover
+import Mathlib.Order.Disjointed
+import Mathlib.Topology.ContinuousMap.Algebra
+
+/-!
+# Uniform approximation by locally constant functions
+
+We show that if `X` is a profinite space, and `V` is a uniform space, then any continuous function
+`X → V` can be uniformly approximated by locally constant functions.
+-/
+
+open UniformSpace Set
+
+open TopologicalSpace (Opens Clopens IsOpenCover)
+
+open scoped Uniformity Function
+
+namespace ContinuousMap
+
+variable {X Y V : Type*}
+  [TopologicalSpace X] [TotallyDisconnectedSpace X] [T2Space X] [CompactSpace X]
+  [UniformSpace V] (f : C(X, V)) {S : Set (V × V)}
+
+/--
+For any continuous function on `X` and `S ∈ 𝓤 V`, there exists a finite cover of `X` by
```

Comment:
I confess I have always thought this to be true without any uniformity assumption on `V`. This is certainly not the case for your proof, and I might very well be wrong: do you have a counterexample?

Reply:
I imagined that it worked under the simple assumption that `S` was symmetric, but I am probably wrong (and since I can probably imagine the applications you have in mind, utterly useless).

Reply:
You are completely right! It is actually true more generally for any subset of `V × V` which is open and contains the diagonal. This generalisation actually follows from the statement here: the range of `f` has to be contained in a compact subset of `V`, so WLOG V is itself compact, and in a compact Hausdorff space the neighbourhoods of the diagonal form a uniformity (the unique uniformity compatible with the topology). 

I will reimplement it in this wider generality (and move it to another location in the hierarchy!).

Reply:
I have now generalized the result so that `V` can be any topological space. 

I tried giving an independent proof (not via uniform spaces), but there is a fiddly step – to show that if `S` is any nhd of the diagonal then there is some smaller nhd `T` of the diagonal such that `T ∘ T ⊆ S` (composition as relations) – and this is actually the main ingredient in the proof that in any compact T2 space the nhds of the diagonal are a uniformity (`uniformSpaceOfCompactT2`). So, rather than duplicating this proof, it is simpler to just quote it. Hence I did not move the result outside the `UniformSpace` hierarchy.

Reply:
It might be me, but I do not see your changes (nor this one about generalising `V` to be any topological space, nor the renaming of the lemma that you commented with a 👍 ): are you sure you `push`ed?

Reply:
You're right, I forgot. But since then I realised a better way of doing it, which I am now working on. So I have set it back to "WIP" for now.